types/field_type.go

```diff
@@ -310,10 +310,19 @@ func MergeFieldType(a byte, b byte) byte {
 }
 
 // mergeTypeFlag merges two MySQL type flag to a new one
-// currently only NotNullFlag is checked
-// todo more flag need to be checked, for example: UnsignedFlag
+// currently only NotNullFlag and UnsignedFlag are checked
+// todo more flag need to be checked
 func mergeTypeFlag(a, b uint) uint {
-	return a & (b&mysql.NotNullFlag | ^mysql.NotNullFlag)
+	c := a
+	if !mysql.HasNotNullFlag(b) {
+		c &= ^mysql.NotNullFlag
+	}
+
+	if !mysql.HasUnsignedFlag(b) {
+		c &= ^mysql.UnsignedFlag
+	}
+
+	return c
```

Comment:
I think the unsigned flag can not be simply set or unset, because the range of tp is changed when the unsigned flag will be changed.  When merge `unsigned int32` and `signed int32`, the best fit type should be `unsigned int64`, and things get more complex when merge `unsigned int64` and `signed int64`, as far as I know, the latest mysql return  `decimal` type, so simply set/unset the unsigned flag many cause some unknown issues, this is the reason why I does not  check unsigned flag in the last pr. In order to handle the unsigned flag, maybe you need to make some changes in `AggFieldType`.

Reply:
@windtalker Right, for `case when` MySQL 8 will use `DECIMAL(20, 0)` for `signed int64` and `unsigned int64`. I'll give a look at the detailed logic.